### PR TITLE
Better ordering in generated dumps

### PR DESCRIPTION
**Pull request by timmytofu**
Would this order not be better for a dump file? Nodes are imported first, without constraints or indices slowing things down, then constraints and indices are added so the relationship creation can use them (maybe, if they're ready)?

The only downside I can think of is if someone is importing into an already populated database and one of the nodes conflicts with a constraint, you'll wait and see the error after a bit of work has already been done. That seems like a rare situation, though.

### CODE REVIEW

**Comment by timmytofu** [comment]
I can't see the error, but I could well have borked a test, I couldn't actually get the tests running in a reasonable amount of time.

Related neo4j#2491 neo4j#2625

**Comment by jotomo** [comment]
@timmytofu The build's most likely failing because you're not on the CI's whitelist, so he refuses to do the build.

What kind of speedups have you measured?

**Comment by jakewins** [comment]
@timmytofu +1, what @jotomo said. Would you mind signing the CLA so that we're allowed to include your contributions? It's a quick email, instructions here: http://neo4j.com/docs/stable/cla.html

As per the contents of the PR - @jexp @boggle y'all got any feedback on this? I'm not familiar with the dump code..

**Comment by jexp** [comment]
It's an interesting suggestion but that only holds true for imports in empty databases, if the database is already populated the constraints have to go in first. Otherwise you might end up with duplicates.

I wanted to rework the dump code anyway to be more sensible. Which is now possible but was not back then.

**Comment by boggle** [comment]
In terms of performance, creating indices last might make more sense as the initial index population is very fast (background store scan) vs many additional index writes per created node. Would be great to measure before merging this.  I agree with Michael that constraints should go first to check against duplicates though.

**Comment by timmytofu** [comment]
@jexp @boggle I mention the empty vs. populated database bit in the second half of the original message. To note, though, even putting the constraints at the top doesn't solve the problem, as if the constraint is being imported into an already populated database there might already be data in violation. Importing constraints into a populated database is inherently unpredictable, hence I still think constraints should be deferred if it does indeed increase speed - throw the weird cases under the bus to speed up the common case. Since it all happens within a transaction, neither way should result in corrupted data, one just fails more quickly by making all imports run more slowly.

On that matter I haven't been able to test it for real speed increases - I tried using our data (~1mil nodes, ~2.5mil relationships, nine distinct labels, fifteen relationship types, several indices and unique constraints) and the current dump output, but it ran for hours and hours and eventually I just gave up. We usually do our importing by copying the `data/` directory, though a reasonably performant CQL dump would be useful for porting data across versions (which we'd want to do since we can't import from CSV in 2.2.1, see #4434).

TL/DR this was just kind of a lark/brainfart. I've sent in the CLA, though, so take it for whatever it's worth.

**Comment by timmytofu** [comment]
Actually, cla (@t) neotechnology dot cöm bounced back at me.

**Comment by jakewins** [comment]
@timmytofu did you email cla@neotechnology.com, or the literal cla(@t)neotechnology.cöm? If should be the former, I think @jexp checked this earlier today, the former should work. Please let us know if it doesn't!

**Comment by timmytofu** [comment]
I used the proper email addy:

cla

**Comment by jakewins** [comment]
Sorry for dropping the ball on this, I've had GH notifications set wrong.

Ping @jexp @simpsonjulian, have there been issues with the CLA email?

**Comment by simpsonjulian** [comment]
@jakewins we're getting CLA email; that one looks like it failed to be relayed on it's way to us.

**Comment by jakewins** [comment]
@boggle @jexp will we merge this in? If there's performance concerns with this approach, perhaps best to close this PR.

**Comment by boggle** [comment]
@timmytofu @jexp @jakewins Thanks for the effort and for measuring the impact. Another reason for putting constraints first would be to prevent importing data if it would violate a later constraint and also roll back importing if the data is not in the right shape. Then again, someone might want to construct a result in order to fix up a database so that the constraint holds, in which case putting constraints last seems more desirable.  It seems to me that this whole area needs more thinking on our part, perhaps the order of needs to be made configurable. Closing for now and happy to re-open if there's a strong argument why some order is always preferable.

**Comment by jotomo** [comment]
FYI: I created a meta issue for all things related to dumps; I've added this issue and linked here. Issue in question in #2625
